velox/expression/SpecialFormRewrites.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <set>
+
+#include "velox/expression/ExprConstants.h"
+#include "velox/expression/ExprOptimizer.h"
+#include "velox/expression/ExprUtils.h"
+#include "velox/expression/SpecialFormRewrites.h"
+
+namespace facebook::velox::expression {
+
+core::TypedExprPtr rewriteConjunctExpression(const core::TypedExprPtr& expr) {
+  if (!utils::isCall(expr, kAnd) && !utils::isCall(expr, kOr)) {
+    return nullptr;
+  }
+
+  const auto* callExpr = expr->asUnchecked<core::CallTypedExpr>();
+  const bool isAnd = (callExpr->name() == kAnd) ? true : false;
+  // If all inputs are `AND` or `OR` then we can flatten the inputs to a vector
+  // before further optimizing.
+  auto canFlatten = utils::allInputTypesEquivalent(expr);
+  std::vector<core::TypedExprPtr> flat;
+  if (canFlatten) {
+    utils::flattenInput(expr, callExpr->name(), flat);
+  }
+
+  const auto& inputsToOptimize = canFlatten ? flat : expr->inputs();
+  bool allInputsConstant = true;
+  std::vector<core::TypedExprPtr> optimizedInputs;
+  core::TypedExprPtr nullInput = nullptr;
+
+  for (const auto& inputExpr : inputsToOptimize) {
+    const auto predicateEvalResult = utils::evalConstantPredicate(inputExpr);
+    switch (predicateEvalResult) {
+      case utils::PredicateEvalResult::kNull: {
+        if (nullInput == nullptr) {
+          nullInput = inputExpr;
+          optimizedInputs.push_back(nullInput);
+        }
+        break;
+      }
+      case utils::PredicateEvalResult::kTrue: {
+        if (!isAnd) {
+          // OR (.., true, ..) -> true
+          return inputExpr;
+        }
+        break;
+      }
+      case utils::PredicateEvalResult::kFalse: {
+        if (isAnd) {
+          // AND (.., false, ..) -> false
+          return inputExpr;
+        }
+        break;
+      }
+      case utils::PredicateEvalResult::kNotConstantBool: {
+        allInputsConstant = false;
+        optimizedInputs.push_back(inputExpr);
+        break;
+      }
+    }
+  }
+
+  if (allInputsConstant && nullInput != nullptr) {
+    return nullInput;
+  }
+  if (optimizedInputs.empty()) {
```

Comment:
Is this possible? When would this happen?

Reply:
I was thinking of cases like `true and true`.